src/drivers/intel/cavs/sue-iomux.c

```diff
@@ -0,0 +1,120 @@
+ *     documentation and/or other materials provided with the distribution.
+ *   * Neither the name of the Intel Corporation nor the
+ *     names of its contributors may be used to endorse or promote products
+ *     derived from this software without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE
+ * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ *
+ * Author: Guennadi Liakhovetski <guennadi.liakhovetski@linux.intel.com>
+ */
+
+#include <errno.h>
+#include <stddef.h>
+#include <stdint.h>
+
+#include <sof/io.h>
+#include <sof/iomux.h>
+
+#include <platform/memory.h>
+
+#define SUE_IOMUX_OFFSET(x)	((x) + EXT_CTRL_BASE)
+#define SUE_IOMUX_CTL0_REG	SUE_IOMUX_OFFSET(0x30)
+#define SUE_IOMUX_CTL1_REG	SUE_IOMUX_OFFSET(0x34)
+
+#define SUE_PIN_NUM 25
+
+struct iomux {
+	/* 0 -- not configured as GPIO; 1 -- configured as GPIO */
+	uint8_t	pin_state[SUE_PIN_NUM];
+};
+
+static struct iomux sue_iomux;
+
+/* Configure as GPIO */
+#define SUE_IOMUX_MODE_GPIO	1
+
+int iomux_configure(struct iomux *iomux, unsigned int port,
+		    enum iomux_function fn)
+{
+	uint32_t shift;
+
+	if (fn != IOMUX_CONFIG_GPIO)
+		return -EINVAL;
+
+	switch (port) {
+	case 0 ... 7:
+		shift = port << 1;
+		io_reg_update_bits(SUE_IOMUX_CTL1_REG, 3 << shift,
+				   SUE_IOMUX_MODE_GPIO << shift);
+		break;
+	case 8:
+		io_reg_update_bits(SUE_IOMUX_CTL1_REG, 1 << 16,
+				   SUE_IOMUX_MODE_GPIO << 16);
+		break;
+	case 9 ... 12:
+		io_reg_update_bits(SUE_IOMUX_CTL0_REG, 1 << 11,
+				   SUE_IOMUX_MODE_GPIO << 11);
+		break;
+	case 13:
+		io_reg_update_bits(SUE_IOMUX_CTL0_REG, 1 << 0,
+				   SUE_IOMUX_MODE_GPIO << 0);
+		break;
+	case 14:
+		io_reg_update_bits(SUE_IOMUX_CTL0_REG, 1 << 1,
+				   SUE_IOMUX_MODE_GPIO << 1);
+		break;
+	case 15 ... 18:
+		io_reg_update_bits(SUE_IOMUX_CTL0_REG, 1 << 9,
+				   SUE_IOMUX_MODE_GPIO << 9);
+		break;
+	case 19 ... 22:
+		io_reg_update_bits(SUE_IOMUX_CTL0_REG, 1 << 10,
+				   SUE_IOMUX_MODE_GPIO << 10);
+		break;
+	case 23:
+	case 24:
+		io_reg_update_bits(SUE_IOMUX_CTL0_REG, 1 << 16,
+				   SUE_IOMUX_MODE_GPIO << 16);
+		break;
+	case 25:
+		io_reg_update_bits(SUE_IOMUX_CTL0_REG, 1 << 26,
+				   SUE_IOMUX_MODE_GPIO << 26);
+		break;
+	default:
+		return -EINVAL;
```

Comment:
a trace error would be good here.

Reply:
...actually all this is wrong, let me rework it. I found documentation that explains what is going on here. It was actually also clear from the code. We have two instances of the IO MUX above - 0 and 1. Some GPIOs are controlled over 0, others over 1. The above function hard-codes GPIO <-> IO MUX relationship instead of providing a reusable driver. Should I so far keep that the simple way without marking configured pins as busy and allowing anyone to reconfigure or do we want a proper resource management for them too? We can add that later...

Reply:
If you need new trace class, please add for SPI, etc